pkg/registry/core/service/storage/storage_test.go

```diff
@@ -5580,3 +5592,81 @@ func TestCreateDryRun(t *testing.T) {
 		})
 	}
 }
+
+// Prove that a dry-run delete doesn't actually deallocate IPs or ports.
+func TestDeleteDryRun(t *testing.T) {
+	testCases := []struct {
+		name            string
+		enableDualStack bool
+		svc             *api.Service
+	}{{
+		name:            "gate:off",
+		enableDualStack: false,
+		svc: svctest.MakeService("foo",
+			svctest.SetTypeLoadBalancer,
+			svctest.SetExternalTrafficPolicy(api.ServiceExternalTrafficPolicyTypeLocal)),
+	}, {
+		name:            "gate:on",
+		enableDualStack: true,
+		svc: svctest.MakeService("foo",
+			svctest.SetTypeLoadBalancer,
+			svctest.SetExternalTrafficPolicy(api.ServiceExternalTrafficPolicyTypeLocal)),
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, tc.enableDualStack)()
+
+			families := []api.IPFamily{api.IPv4Protocol}
+			if tc.enableDualStack {
+				families = append(families, api.IPv6Protocol)
+			}
+
+			storage, _, server := newStorage(t, families)
+			defer server.Terminate(t)
+			defer storage.Store.DestroyFunc()
+
+			ctx := genericapirequest.NewDefaultContext()
+			createdObj, err := storage.Create(ctx, tc.svc, rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("unexpected error creating service: %v", err)
+			}
+			createdSvc := createdObj.(*api.Service)
+
+			// Ensure IPs and ports were allocated
+			for i, fam := range createdSvc.Spec.IPFamilies {
+				if !ipIsAllocated(t, storage.alloc.serviceIPAllocatorsByFamily[fam], createdSvc.Spec.ClusterIPs[i]) {
+					t.Errorf("expected IP to be allocated: %v", createdSvc.Spec.ClusterIPs[i])
+				}
+			}
+			for _, p := range createdSvc.Spec.Ports {
+				if !portIsAllocated(t, storage.alloc.serviceNodePorts, p.NodePort) {
+					t.Errorf("expected port to be allocated: %v", p.NodePort)
+				}
+			}
+			if !portIsAllocated(t, storage.alloc.serviceNodePorts, createdSvc.Spec.HealthCheckNodePort) {
+				t.Errorf("expected port to be allocated: %v", createdSvc.Spec.HealthCheckNodePort)
+			}
+
+			_, _, err = storage.Delete(ctx, tc.svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{DryRun: []string{metav1.DryRunAll}})
+			if err != nil {
+				t.Fatalf("unexpected error deleting service: %v", err)
+			}
+
+			// Ensure they are still allocated.
+			for i, fam := range createdSvc.Spec.IPFamilies {
+				if !ipIsAllocated(t, storage.alloc.serviceIPAllocatorsByFamily[fam], createdSvc.Spec.ClusterIPs[i]) {
+					t.Errorf("expected IP to still be allocated: %v", createdSvc.Spec.ClusterIPs[i])
+				}
+			}
+			for _, p := range createdSvc.Spec.Ports {
+				if !portIsAllocated(t, storage.alloc.serviceNodePorts, p.NodePort) {
+					t.Errorf("expected port to still be allocated: %v", p.NodePort)
```

Comment:
Commit `De-layer Service REST: Delete`:
`%+v` ? (similarly other struct types)

Reply:
ACK, will try to use %q for IPs and %d for ports.